src/logging/writers/ascii/Ascii.cc

```diff
@@ -16,6 +28,146 @@ using namespace threading;
 
+static constexpr auto shadow_file_prefix = ".shadow.";
+
+/**
+ * Information about an leftover log file: that is, one that a previous
+ * process was in the middle of writing, but never completed a rotation
+ * for whatever reason (prematurely crashed/killed).
+ */
+struct LeftoverLog {
+	/*
+	 * Name of leftover log, relative to working dir.
+	 */
+	std::string filename;
+
+	/*
+	 * File extension of the leftover log (e.g. ".log").
+	 */
+	std::string extension;
+
+	/*
+	 * Name of shadow file associated with the log.
+	 * The shadow file's existence is what indicates the presence of
+	 * an "leftover log" and may contain the name of a postprocessing
+	 * function that's supposed to be called after rotating (only
+	 * named if that function differs from the default).  Upon
+	 * completing a rotation, the shadow file can be deleted.
+	 */
+	std::string shadow_filename;
+
+	/**
+	 * Name of a function to call to postprocess the log file after
+	 * rotating.
+	 */
+	std::string post_proc_func;
+
+	/**
+	 * The time at which the shadow file was created.  This is used
+	 * as the log file's "opening time" for rotation purposes.
+	 */
+	time_t open_time;
+
+	/**
+	 * Time of the log file's last modification.  This is used
+	 * as the log file's "closing time" for rotation purposes.
+	 */
+	time_t close_time;
+
+	/**
+	 * Set the an error message explaining any error that happened while
+	 * trying to parse the shadow file and construct an object.
+	 */
+	std::string error;
+
+	/**
+	 * Return the name of the log without file extension (this is
+	 * what's called "path" in other logging framework parlance).
+	 * E.g. the name of "conn.log" is just "conn".
+	 */
+	std::string Name() const
+		{ return filename.substr(0, filename.size() - extension.size()); }
+
+	/**
+	 * Deletes the shadow file and returns whether it succeeded.
+	 */
+	bool DeleteShadow() const
+		{ return unlink(shadow_filename.data()) == 0; }
+};
+
+static std::optional<LeftoverLog> parse_shadow_log(const std::string& fname)
+	{
+	char errbuf[512];
+	auto sfname = shadow_file_prefix + fname;
+
+	LeftoverLog rval;
+	rval.filename = fname;
+	rval.shadow_filename = std::move(sfname);
+
+	auto sf_stream = fopen(rval.shadow_filename.data(), "r");
+
+	if ( ! sf_stream )
+		{
+		bro_strerror_r(errno, errbuf, sizeof(errbuf));
+		rval.error = "Failed to open " + rval.shadow_filename + ": " + errbuf;
+		return rval;
+		}
+
+	fseek(sf_stream, 0, SEEK_END);
+	auto sf_len = ftell(sf_stream);
+	fseek(sf_stream, 0, SEEK_SET);
+
+	auto sf_content = std::make_unique<char[]>(sf_len);
+	auto bytes_read = fread(sf_content.get(), 1, sf_len, sf_stream);
+
+	if ( bytes_read != static_cast<size_t>(sf_len) )
+		{
+		rval.error = "Failed to read contents of " + rval.shadow_filename;
+		return rval;
+		}
+
+	fclose(sf_stream);
+	std::string_view sf_view(sf_content.get(), sf_len);
+	auto sf_lines = tokenize_string(sf_view, '\n');
+
+	if ( sf_lines.size() < 2 )
+		{
+		snprintf(errbuf, sizeof(errbuf),
```

Comment:
Just use `fmt`?

Reply:
Sure (guessing this originally lived within the thread where that wasn't safe to do).